sdk/servicebus/service-bus/src/serviceBusClient.ts

```diff
@@ -56,23 +57,87 @@ export class ServiceBusClient {
    */
   private _context: ConnectionContext;
 
+  /**
+   * Creates a ServiceBusClient for the Service Bus Namespace represented in the given connection
+   * string.
+   * @param connectionString - Connection string of the form
+   * 'Endpoint=sb://my-servicebus-namespace.servicebus.windows.net/;SharedAccessKeyName=my-SA-name;SharedAccessKey=my-SA-key'
+   * @param options Options to control ways to interact with the
+   * Service Bus Namespace.
+   * @returns ServiceBusClient
+   */
+  constructor(connectionString: string, options?: ServiceBusClientOptions);
+
   /**
    * Instantiates a ServiceBusClient to interact with a Service Bus Namespace.
    *
    * @constructor
-   * @param {ConnectionConfig} config - The connection configuration needed to connect to the
-   * Service Bus Namespace.
-   * @param {TokenCredential} [tokenCredential] - SharedKeyCredential object or your 
-   * credential that implements the TokenCredential interface.
-   * @param {ServiceBusClientOptions} - Options to control ways to interact with the Service Bus
+   * @param host - The host name for the Service Bus namespace. This is likely to be similar to
+   * <yournamespace>.servicebus.windows.net
+   * @param credential - credential that implements the TokenCredential interface.
+   * @param options - Options to control ways to interact with the Service Bus
    * Namespace.
    */
-  private constructor(
-    config: ConnectionConfig,
-    credential: SharedKeyCredential | TokenCredential,
+  constructor(host: string, credential: TokenCredential, options?: ServiceBusClientOptions);
+  constructor(
+    hostOrConnectionString: string,
+    credentialOrServiceBusClientOptions?: TokenCredential | ServiceBusClientOptions,
     options?: ServiceBusClientOptions
   ) {
-    if (!options) options = {};
+    let config;
+    let credential;
+    let connectionString;
+
+    if (!options) {
+      options = {};
+    }
+
+    if (
+      hostOrConnectionString == undefined ||
+      hostOrConnectionString.toString == undefined ||
+      hostOrConnectionString.toString().trim() == ""
+    ) {
+      throw new Error(
+        "Input parameter of host or connection string must be defined and coercible to string."
+      );
+    }
+    hostOrConnectionString = hostOrConnectionString.toString();
+
+    let isConnectionStringConstructorInvoked = true;
+    if (hostOrConnectionString.indexOf("=") < 0) {
+      isConnectionStringConstructorInvoked = false;
+    }
```

Comment:
The above checks are not needed. `ConnectionConfig.create()` ensures that `String()` is called on the given connection string. If the connection string thus stringified doesnt meet the requirements of a connection string, then appropriate errors are thrown from `ConnectionConfig.create()`.

I would suggest the below for handling the overloads

```javascript
if (!isTokenCredential(credentialOrServiceBusClientOptions)) {
    // assume that first overload is used & create the ConnectionConfig
} else {
    // assume that second overload is used & create the ConnectionConfig
}
// create the connection context
```

Reply:
The reason for converting to String is because if it's a hostname then we use a .endsWith() check in here

Reply:
You should still be able to use `hostOrConnectionString.endsWith()`, TS should not complain as the type is already set to `string` in the constructor signature.

If you want to avoid the case of runtime error for when a bad type is passed, then you can use `hostOrConnectionString = String(hostOrConnectionString)`. Since then endsWith() check is needed only in the case of TokenCredential, you can move this inside the if/else block as well.

Reply:
Any reason why we want to convert it to type `String` and not use `.toString()`? If i remember correctly, I think @bterlson had once discussed, advised about not doing so and using `string` where possible.
And yes, the conversion was to help with both compile and run time scenarios.

Reply:
My intention here is to simplify the logic to handle the constructor overloads. 
Currently, we are adding multiple lines of code to ensure that `hostOrConnectionString` is indeed a string and that it has `=` in it. We can avoid the need for this and the future tax of maintaining them as well.

For example, when the constructor overload is using connection string, we shouldnt need to do any checks on `hostOrConnectionString` as that work is offloaded to the downstream `ConnectionConfig.create()`.

When the constructor overload for credential is being used, us using String() will ensure that if user had provided an object with .toString() defined, then it gets called and used correctly. If the passed object does not have .toString() on it, then the connection to the service would fail eventually with the service providing an error message hinting that the provided host name is not valid.

Reply:
I think there's a trade off with quality of expected user experience here possibly - for instance, if a user is passing in the hostname correctly and receives a connection string malformed error, wouldn't that be confusing / inappropriate?

Reply:
For above, let's say the inputs are passed along with a malformed / incorrect / unsupported type of credentials.
Current changes address these cases. By offloading the check to service or core-amqp to expect a connection string - would make it not possible to do so.

Reply:
Updated to use a different signature for the credential based constructor - should likely help with both simplification/maintenance and user ex concerns.

Reply:
I'm just catching up here and I might have missed some things as changes have been made.
1st, I remarked on this in the API file but I do think we want to keep `connectionString` and `host` as the first parameter in their respective constructors to be consistent with our other SDKs/guidelines.

I think the original comment was about the checks being done on `hostOrConnectionString`.
```
if (
      hostOrConnectionString == undefined ||
      hostOrConnectionString.toString == undefined ||
      hostOrConnectionString.toString().trim() == ""
    ) {
      throw new Error(
        "Input parameter of host or connection string must be defined and coercible to string."
      );
    }
```

I think @ramya-rao-a 's had 2 points:
1. We can ensure `hostOrConnectionString` is a string by wrapping it in `String`:  `String(hostOrConnectionString)`
Note that this is different from `new String(hostOrConnectionString)`, which is not what you want (and what I think @bterlson would warn to stay away from).
2. There are other places where `hostOrConnectionString` is validated so this check isn't needed.

The `EventHubClient` constructor shows both of these aspects off (only linking to one line here but look at the rest of the function): https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/src/impl/eventHubClient.ts#L339

Event Hubs is a bit more complicated since it has 3 overloads instead of 2, so service bus would be even simpler.

I think the Event Hubs constructor would throw helpful errors for the cases that were being checked before, but if there's something being missed let me know!